crypto/ec/ec_evp_lib.c

```diff
@@ -0,0 +1,417 @@
+        return -2;
+    }
+
+    *p++ = OSSL_PARAM_construct_int(OSSL_EXCHANGE_PARAM_EC_ECDH_COFACTOR_MODE,
+                                    &ret);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (!EVP_PKEY_CTX_get_params(ctx, params))
+        return -1;
+
+    if (ret < 0 || ret > 1) {
+        /*
+         * The provider should return either 0 or 1, any other value is a
+         * provider error.
+         */
+        return -1;
+    }
+
+    return ret;
+}
+
+int EVP_PKEY_CTX_set_ecdh_kdf_type(EVP_PKEY_CTX *ctx, int kdf)
+{
+    int ret;
+    OSSL_PARAM params[2], *p = params;
+
+    if ( 1 != (ret = evp_pkey_ctx_getset_ecdh_param_checks(ctx)))
+        return ret;
+
+    if (kdf != EVP_PKEY_ECDH_KDF_NONE && kdf != EVP_PKEY_ECDH_KDF_X9_63)
+        return -2;
+
+    /* TODO(3.0): Remove this eventually when no more legacy */
+    if (ctx->op.kex.exchprovctx == NULL)
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_TYPE, kdf, NULL);
+
+    /* Check the ctx actually understands this parameter */
+    if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_settable_params(ctx),
+                OSSL_EXCHANGE_PARAM_KDF_TYPE) == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
+
+    /* TODO(romen) : magic numbers or strings toward provider? */
+    *p++ = OSSL_PARAM_construct_int(OSSL_EXCHANGE_PARAM_KDF_TYPE,
+                                    &kdf);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (!EVP_PKEY_CTX_set_params(ctx, params))
+        return 0;
+
+    return 1;
+}
+
+/* TODO(3.0): should constify? */
+int EVP_PKEY_CTX_get_ecdh_kdf_type(EVP_PKEY_CTX *ctx)
+{
+    int ret;
+    OSSL_PARAM params[2], *p = params;
+
+    if ( 1 != (ret = evp_pkey_ctx_getset_ecdh_param_checks(ctx)))
+        return ret;
+
+    /* TODO(3.0): Remove this eventually when no more legacy */
+    if (ctx->op.kex.exchprovctx == NULL)
+        return EVP_PKEY_CTX_ctrl(ctx, EVP_PKEY_EC,
+                                 EVP_PKEY_OP_DERIVE,
+                                 EVP_PKEY_CTRL_EC_KDF_TYPE, -2, NULL);
+
+    /* Check the ctx actually understands this parameter */
+    if (OSSL_PARAM_locate_const(EVP_PKEY_CTX_gettable_params(ctx),
+                OSSL_EXCHANGE_PARAM_KDF_TYPE) == NULL) {
+        ERR_raise(ERR_LIB_EVP, EVP_R_COMMAND_NOT_SUPPORTED);
+        /* Uses the same return values as EVP_PKEY_CTX_ctrl */
+        return -2;
+    }
+
+    *p++ = OSSL_PARAM_construct_int(OSSL_EXCHANGE_PARAM_KDF_TYPE,
+                                    &ret);
+    *p++ = OSSL_PARAM_construct_end();
+
+    if (!EVP_PKEY_CTX_get_params(ctx, params))
+        return -1;
+
+    /* TODO(romen) : magic numbers or strings toward provider? */
+    if (ret != EVP_PKEY_ECDH_KDF_NONE && ret != EVP_PKEY_ECDH_KDF_X9_63)
+        return -2;
```

Comment:
@mattcaswell @levitte 

What is the best approach here? we have to expose magic numbers to the EVP layer, but maybe to the provider side we could use strings?

Consider that this same parameter could be used for the DH provider implementation (that currently does not include the KDF stuff): we could probably avoid potentially clashing magic numbers and use strings towards the provider.

What would be the best approach to use a predefined set of strings as the value for this param?

Reply:
We already have a set of "Known KDF names" in include/openssl/core_names.h (including one for `OSSL_KDF_NAME_X963KDF`). Perhaps use those?

Reply:
I'll try that!

Reply:
Done